### PR TITLE
[DLSP21] 12-1

### PR DESCRIPTION
**Pull request by rohithmukku**
Authors/Contributors: Fanzeng Xia (first half),  Rohith Mukku (second half)

Please check the PR and suggest if any changes are required.

### CODE REVIEW

File: docs/en/week12/12-1.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
+---
+lecturer: Marc'Aurelio Ranzato
+title: Low Resource Machine Translation (Part 1)
+authors: Fanzeng Xia, Rohith Mukku
+date: 21 April 2021
+---
+
+# Low Resource Machine Translation
```
Comment (Atcold) [inline docs/en/week12/12-1.md:8]:
```suggestion

```

File: docs/en/week12/12-1.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+<center>
+<img src="{{site.baseurl}}/images/week12/12-1/figure2.png" style="background-color:#DCDCDC;"><br>
+<b>Figure 2:</b> A parallel dataset consist of Italian and English
+</center>
+ 
+### Train NMT
+The standard way of training NMT is by using maximum likelihood. Give a source sentence $x$ and the model parameters $\theta$, we will seek to maximize the log likelihood of the joint probability of all the ordered sequence of tokens in the target sentence $y$, as illustrated by the equation below:
+
+$$ - \log\ p(y|x; \theta) = - \sum_{j=1}^{n} log\ p(y_j|y_{j-1},...y_1,x; \theta) $$
```
Comment (Atcold) [inline docs/en/week12/12-1.md:29]: `log` needs `\` before.
The conditional bar is `\mid`.
Vectors need `\vect{}`.

Here and elsewhere.

File: docs/en/week12/12-1.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+It is used as a score to indicate how likely the target sentence acutally to be a translation from the source sentence. A minus sign is added in front of the equation to turn this into a minimization problem. 
+
+#### Language Model
+In language modelling, given a token $X_t$ at time $t$, we are trying to predict the next word $Y_t$. As shown in the figure below, hidden state $Z_t$ is generated by a RNN/CNN/Transformer block depending on the input vector $X_t$ and the previous hidden state $Z_{t-1}$.
```
Comment (Atcold) [inline docs/en/week12/12-1.md:34]: Why are we using capitals now?

File: docs/en/week12/12-1.md [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+Each potential choice (the number of words in vocabulary) has a probability score. At every step, beam search select the top K scoring among all the branches (maintain queue with k top scoring paths), then expand each of them and retain the top k scoring paths. As illustrated by the example below:
+
+<center>
+<img src="{{site.baseurl}}/images/week12/12-1/figure7.png" style="background-color:#DCDCDC;"><br>
+<b>Figure 7:</b> Beam Search
+</center>
+
+
+Let k = 2
```
Comment (Atcold) [inline docs/en/week12/12-1.md:92]: `$` missing.

File: docs/en/week12/12-1.md [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+### NMT Training & Inference Summary
+
+Training: predict one target token at the time and minimize cross-entropy loss
+
+Inference: find the most likely target sentence (approximately) using beam search
+
+Evaluation: compute BLEU on hypothesis returned by the inference procedure
+
+$p_n = \frac{\sum_{generated \ sentences} \sum_{ngrams} Clip(Count(ngram matches))}{\sum_{generated \ sentences} \sum_{ngrams} Count(ngram)}$
```
Comment (Atcold) [inline docs/en/week12/12-1.md:113]: Double `$`.
Use `\text{}` for text.

File: docs/en/week12/12-1.md [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+$BLEU = BP \ e^{\sum_{n=1}^{N} \frac{1}{N} log \ p_n}$
```
Comment (Atcold) [inline docs/en/week12/12-1.md:115]: What is `B` and what is `P`?
Use `\exp`.

File: docs/en/week12/12-1.md [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+These sentences are translated and the quality of the translations are determined using the automatic checks and human evaluation. There are several techniques in automatic checks, a model was trained on each language and perplexity was measured. If the perplexity is too high, then we go back to translation stage as shown above. Other checks are transliteration, using Google Translate, etc. There is no criteria or conditions for when this loop is stopped, essentially it depends on the perplexity and any outliers.
+
+<center>
+<img src="{{site.baseurl}}/images/week12/12-1/figure18.png" style="background-color:#DCDCDC;"><br>
+<b>Figure 18:</b> Some examples of English Sinhala translations
+</center>
+
+Figure 18 shows some examples. As we can see, they are not really fluent. One more issue with these is that the wikipedia articles of Sinhala language mostly contain topics of very limited number of domains (like their religion, country, etc.)
+
+Continuation: Model and Analysis stages will be discussed in the following lecture 12.2.
```
Comment (Atcold) [inline docs/en/week12/12-1.md:221]:
```suggestion

```

**Review by Atcold — CHANGES_REQUESTED** [review]
Mostly good. A few minor changes.

**Comment by rohithmukku** [comment]
I have made the changes, please check.

File: docs/en/week12/12-1.md [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+
+<center>
+<img src="{{site.baseurl}}/images/week12/12-1/figure4.png" style="background-color:#DCDCDC;"><br>
+<b>Figure 4:</b> Represent source
+</center>
+
+<br/>
+
+Step 2: Score each source word (attention) 
+Take the dot product between target hidden representation $Z_{t+1}$ and all the embeddings from source sentence. Then use softmax to get a distribution over source tokens (attention).
```
Comment (Atcold) [inline docs/en/week12/12-1.md:54]: I said no capital letter for variables, plz.

File: docs/en/week12/12-1.md [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+
+Evaluation: compute BLEU on hypothesis returned by the inference procedure
+
+We first compute the geometric average of the modified n-gram precisions, $p_n$
+
+$$p_n = \frac{\sum_{\text{generated sentences}} \sum_{\text{ngrams}} \text{Clip(Count(ngram matches))}}{\sum_{\text{generated sentences}} \sum_{\text{ngrams}} \text{Count(ngram)}}$$
+
+Candidate translations longer than their references are already penalized by the modified n-gram precision measure: there is no need to penalize them again. Consequently, we introduce a multiplicative brevity penalty factor, $BP$. With $BP$ in place, a high-scoring candidate translation must now match the reference translations in length, in word choice, and in word order.
+
+$$BLEU = BP \exp ({\sum_{n=1}^{N} \frac{1}{N} \log \ p_n})$$
```
Comment (Atcold) [inline docs/en/week12/12-1.md:116]: Please, don't use multi-letter names, this is not math at all.
`$S_{\text{BLEU}}$` for BLEU score could be an option.
Again, fix the B * P thing.
Parenthesis should have a `\` in front.

Comment (rohithmukku) [reply docs/en/week12/12-1.md:116]: I apologise, this was not my part but I will fix this ASAP.

Changes I am making:
- Changing BLEU score to `s_{\text{BLEU}}` and BP factor to `f_{BP}`
- `\` is not working, so I am using `\left(` and `\right)` instead

If these changes are okay, I will incorporate them.

Comment (Atcold) [reply docs/en/week12/12-1.md:116]: Go ahead.

**Review by Atcold — CHANGES_REQUESTED** [review]
Use `\vect{}` for **all** vectors across the entire document.
Use `$` for k in the text **every time** you use it.

**Comment by rohithmukku** [comment]
I have used `\vect` for all vectors and used `$k$` everywhere. Please check and let me know if there any other changes required.

File: docs/en/week12/12-1.md [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+2. Continue from Life and Today, expand each of them and calculate the path score back to the start symbol, pick the top 2 scoring path ("Like was" and "Like is") 
+3. So on and so for, keep proceeding until hit the end of the sentence
+4. At the very last step, select the highest scoring path
+
+Beam search is a greedy procedure and it is very effective in practice. There is a trade-off between computational cost and approximation error (the larger the value of $k$, the better the approximation and the higher the computational cost).
+
+Issues of Beam Search:
+Beam search will always select the highest scoring path. Thus, the solution will be biased because of such strategy. It doesn't handle uncertainty well. 
+
+Other decoding methods: sampling, top-k sampling, generative and discriminative reranking
```
Comment (Atcold) [inline docs/en/week12/12-1.md:100]: Use `$`…

File: docs/en/week12/12-1.md [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+<center>
+<b>Figures 12, 13:</b> Data collection and Low-resource tasks in ML
+</center>
+
+#### Challenges
+
+Figure 14 depicts some of the challenges encountered in Low Resource Machine Translation tasks.
+
+<center>
+<img src="{{site.baseurl}}/images/week12/12-1/figure14.png" style="background-color:#DCDCDC;"><br>
```
Comment (Atcold) [inline docs/en/week12/12-1.md:175]: This is text. Please, use text.

**Review by Atcold — CHANGES_REQUESTED** [review]
Almost there!

File: docs/en/week12/12-1.md (hunk 10 above)
Comment (rohithmukku) [reply docs/en/week12/12-1.md:100]: I thought top-k is one word. Okay, will fix that.

File: docs/en/week12/12-1.md (hunk 11 above)
Comment (rohithmukku) [reply docs/en/week12/12-1.md:175]: I don't understand, the text is not enclosed in `$$`. What needs to be done?

Comment (Atcold) [reply docs/en/week12/12-1.md:175]: The figure is not an image, it's text.
Please, type out the text.

File: docs/en/week12/12-1.md (hunk 10 above)
Comment (Atcold) [reply docs/en/week12/12-1.md:100]: `top-$k$`, yup.

**Comment by Atcold** [comment]
Thanks!

**Pull request merged**